filter.go

```diff
@@ -8,60 +8,126 @@ import (
 	manet "github.com/multiformats/go-multiaddr-net"
 )
 
+type filterEntry struct {
+	f      *net.IPNet
+	reject bool
+}
+
+// Filters is a structure representing a collection of allow/deny
+// net.IPNet filters, together with the RejectByDefault flag, which
+// represents the default filter policy.
+//
+// Note that the last policy added to the Filters is authoritative.
 type Filters struct {
-	mu      sync.RWMutex
-	filters map[string]*net.IPNet
+	mu              sync.RWMutex
+	RejectByDefault bool
+	filters         []*filterEntry
 }
 
+// NewFilters constructs and returns a new set of net.IPNet filters.
+// By default, the new filter rejects no addresses.
 func NewFilters() *Filters {
 	return &Filters{
-		filters: make(map[string]*net.IPNet),
+		RejectByDefault: false,
+		filters:         make([]*filterEntry, 0),
 	}
 }
 
+// AddDialFilter adds a reject rule to the given Filters.  Hosts
+// matching the given net.IPNet filter will be rejected, unless
+// another rule is added which states that they should be accepted.
+//
+// No effort is made to prevent duplication of filters, or to simplify
+// the filters list.
 func (fs *Filters) AddDialFilter(f *net.IPNet) {
 	fs.mu.Lock()
 	defer fs.mu.Unlock()
-	fs.filters[f.String()] = f
+	fs.filters = append(fs.filters, &filterEntry{f: f, reject: true})
+}
+
+// AddAllowFilter adds an accept rule to the given Filters. Hosts
+// matching the given net.IPNet filter will be accepted, unless
+// another policy is added which states that they should be rejected.
+//
+// No effort is made to prevent duplication of filters, or to simplify
+// the filters list.
+func (fs *Filters) AddAllowFilter(f *net.IPNet) {
+	fs.mu.Lock()
+	defer fs.mu.Unlock()
+	fs.filters = append(fs.filters, &filterEntry{f: f, reject: false})
 }
 
+// AddrBlocked parses a ma.Multiaddr, and if it can get a valid netip
```

Comment:
`and, if it can get a valid net.IPNet back, applies the Filters,...`

(I'm no native English speaker so I'm not 100% sure but there's something weird with the commas in the original formulation).